ios/RNMobileDeviceManager.m

```diff
@@ -110,33 +125,47 @@ - (NSDictionary *)constantsToExport
 RCT_EXPORT_METHOD(isSingleAppModeEnabled: (RCTPromiseResolveBlock)resolve
                   rejecter:(RCTPromiseRejectBlock)reject)
 {
-    resolve(@(UIAccessibilityIsGuidedAccessEnabled()));
+    dispatch_semaphore_wait(self.asamSem, DISPATCH_TIME_FOREVER);
+    dispatch_async(dispatch_get_main_queue(), ^{
+        resolve(@(UIAccessibilityIsGuidedAccessEnabled()));
+        dispatch_semaphore_signal(self.asamSem);
+    });
 }
 
 RCT_EXPORT_METHOD(isAutonomousSingleAppModeEnabled: (RCTPromiseResolveBlock)resolve
                   rejecter:(RCTPromiseRejectBlock)reject)
 {
-    [MobileDeviceManager isASAMSupported:^(BOOL isSupported){
-        resolve(@((BOOL)(isSupported && UIAccessibilityIsGuidedAccessEnabled())));
+
+    [self isASAMSupported:^(BOOL isSupported){
+        dispatch_async(dispatch_get_main_queue(), ^{
+            resolve(@((BOOL)(isSupported && UIAccessibilityIsGuidedAccessEnabled())));
+        });
     }];
 }
 
 RCT_EXPORT_METHOD(enableAutonomousSingleAppMode: (RCTPromiseResolveBlock)resolve
                   rejecter:(RCTPromiseRejectBlock)reject)
 {
-
-    UIAccessibilityRequestGuidedAccessSession(YES, ^(BOOL didSucceed) {
-        resolve(@(didSucceed));
+    dispatch_semaphore_wait(self.asamSem, DISPATCH_TIME_FOREVER);
+    dispatch_async(dispatch_get_main_queue(), ^{
+        UIAccessibilityRequestGuidedAccessSession(YES, ^(BOOL didSucceed) {
+            resolve(@(didSucceed));
+            dispatch_semaphore_signal(self.asamSem);
```

Comment:
Unlocking should be the first things that happens once completed, otherwise that can lead to a deadlock if it performs a callback, where that callback might wait on the semaphore too